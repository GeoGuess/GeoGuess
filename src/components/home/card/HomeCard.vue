<template>
    <v-card class="map-card" rounded="lg" width="200">
        <v-img
            class="white--text align-end"
            height="140px"
            gradient="rgba(0,0,0,0), rgba(0,0,0,0.8)"
            :src="
                data.imageUrl ||
                    `https://source.unsplash.com/500x230/weekly?${encodeURI(
                        data.nameLocate
                    )}`
            "
        >
            <v-menu v-if="data.custom">
                <template v-slot:activator="{ on, attrs }">
                    <v-btn
                        icon
                        absolute
                        top
                        right
                        dark
                        v-bind="attrs"
                        v-on="on"
                    >
                        <v-icon> mdi-file</v-icon>
                    </v-btn>
                </template>
                <v-list>
                    <v-list-item @click="editMap">
                        <v-list-item-title>
                            {{ $t('Home.HomeCard.edit') }}
                        </v-list-item-title>
                    </v-list-item>
                    <v-list-item @click="deleteMap">
                        <v-list-item-title>
                            {{ $t('Home.HomeCard.delete') }}
                        </v-list-item-title>
                    </v-list-item>
                </v-list>
            </v-menu>

            <v-card-title>
                {{ data.nameLocate }}
            </v-card-title>
        </v-img>
        <v-card-actions class="map-card__actions">
            <v-subheader>{{ $t('Home.HomeCardMode.' + type) }}</v-subheader>
            <v-spacer />
            <HomeCardDialog :data="data" :type="type" />
        </v-card-actions>
    </v-card>
</template>

<script>
import HomeCardDialog from '@/components/home/card/HomeCardDialog';
import { mapActions } from 'vuex';
export default {
    name: 'HomeCard',
    components: {
        HomeCardDialog,
    },
    props: {
        data: Object,
        type: {
            type: String,
            validator: (v) => ['map', 'area'].includes(v),
        },
    },
    methods: {
        ...mapActions(['getListMapsCustoms', 'setMapLoaded']),
        async deleteMap() {
            if (this.type === 'map' && this.data.custom) {
                await this.data.delete();
                this.getListMapsCustoms();
            }
        },
        async editMap() {
            this.setMapLoaded(this.data);
            this.$router.push('/custom');
        },
    },
};
</script>

<style lang="scss">
.map-card__actions {
    padding: 1%;
}
.map-card {
    min-width: 300px;
}
</style>
