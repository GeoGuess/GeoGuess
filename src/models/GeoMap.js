import i18n from '../lang';
import IndexedDBService from '../plugins/IndexedDBService';
import { getLocateString } from '../utils';

export const GeoMapType = {
    Default: 'default',
    Custom: 'custom',
    OSM: 'osm'
};

export class GeoMap {
    constructor() {
        this.geojson = null;
        this.type = GeoMapType.Default;
    }

    get nameLocate() {
        return getLocateString(this, 'name', i18n.locale);
    }

    get descriptionLocate() {
        return getLocateString(this, 'description', i18n.locale);
    }

    get imageSrc() {
        return this.imageUrl ||
            `https://source.unsplash.com/500x230/weekly?${encodeURI(getLocateString(this, 'name', 'en'))}`;
    }
    
    get details(){
        return {
            type: this.type,
            id: this.id,
            name: this.nameLocate,
        };
    }

}

export class GeoMapOSM extends GeoMap {
    constructor(name, osmId, osmType, geojson) {
        super();
        this.name = name;
        this.osmId = osmId;
        this.osmType = osmType;
        this.geojson = geojson;
        this.type = GeoMapType.OSM;
    }

    get details(){
        return {
            type: this.type,
            osmType: this.osmType,
            osmId: this.osmId,
            name: this.name,
        };
    }
}

export class GeoMapCustom extends GeoMap {
    constructor() {
        super();
        this.name = '';
        this.id = undefined;
        this.type = GeoMapType.Custom;
    }
    get nameLocate() {
        return this.name;
    }
    get descriptionLocate() {
        return '';
    }
    async save() {
        if (!this.id) {
            this.dateCreated = new Date();
            this.id = await IndexedDBService.addMap({
                name: this.name,
                dateCreated: this.dateCreated,
                geojson: this.geojson,
            });
        } else {
            this.updateMap = new Date();
            await IndexedDBService.updateMap({
                id: this.id,
                name: this.name,
                dateCreated: this.dateCreated,
                updateMap: this.updateMap,
                geojson: this.geojson,
            });
        }
        return this;
    }
    delete() {
        IndexedDBService.deleteMap(this);
    }
}
