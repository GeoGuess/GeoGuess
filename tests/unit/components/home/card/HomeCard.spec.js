import HomeCard from '@/components/home/card/HomeCard.vue';
import { createLocalVue, shallowMount } from '@vue/test-utils';
import appInit from '../../../testutils/appInit';
import {GeoMapCustom} from '@/models/GeoMap';

const args = appInit(createLocalVue());
const data = {
    name: {
        en: 'Biggest City',
        fr: 'Grande Villes',
    },
    description: {
        fr:
            'Une liste de plus de 40 villes : PARIS, KOBE OSAKA, SEOUL, HOUSTON, BARCELONA, PHILADELPHIA, SANTIAGO, LAGOS, DALLAS, NEW YORK, ISTANBUL, TOKYO, FUKUOKA, LONDON, KUALA LUMPUR, LIMA, HO CHI MINH CITY, MANILA, GUADALAJARA, MADRID, NAGOYA, SINGAPORE, JOHANNESBURG, BELO HORIZONTE, TORONTO, MEXICO CITY, MIAMI, ATLANTA, RIO DE JANEIRO, BUENOS AIRES, SAO PAULO, CHICAGO,…',
        en:
            'List of 40 biggest cities of the world: PARIS, KOBE OSAKA, SEOUL, HOUSTON, BARCELONA, PHILADELPHIA, SANTIAGO, LAGOS, DALLAS, NEW YORK, ISTANBUL, TOKYO, FUKUOKA, LONDON, KUALA LUMPUR, LIMA, HO CHI MINH CITY, MANILA, GUADALAJARA, MADRID, NAGOYA, SINGAPORE, JOHANNESBURG, BELO HORIZONTE, TORONTO, MEXICO CITY, MIAMI, ATLANTA, RIO DE JANEIRO, BUENOS AIRES, SAO PAULO, CHICAGO,…',
    },
    author: 'BilelJegham',
    imageUrl:
        'https://images.unsplash.com/photo-1477959858617-67f85cf4f1df?ixlib=rb-1.2.1&ixid=eyJhcHBfaWQiOjEyMDd9&auto=format&fit=crop&w=300&q=140',
    url: 'https://mapurl.geojson',
};
describe('HomeCard.vue', () => {
    let store;

    it('render', () => {

        const wrapper = shallowMount(HomeCard, {
            ...args,
            store,
            propsData: {
                data,
            },
        });

        expect(wrapper).toMatchSnapshot();
    });
    
    
    it('deleteMap', () => {   
        const map = new GeoMapCustom();
        const spy= jest.spyOn(map,'delete');
        const wrapper = shallowMount(HomeCard, {
            ...args,
            store,
            propsData: {
                data: map,
                type: 'map'
            },
        });

        wrapper.vm.deleteMap();

        expect(spy).toBeCalled();
    });
});
