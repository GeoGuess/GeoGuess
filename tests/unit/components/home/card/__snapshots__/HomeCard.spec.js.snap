// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`HomeCard.vue render 1`] = `
<v-card-stub loaderheight="4" tag="div" width="200" rounded="lg" class="map-card">
  <v-img-stub height="140px" gradient="rgba(0,0,0,0), rgba(0,0,0,0.8)" options="[object Object]" position="center center" src="https://images.unsplash.com/photo-1477959858617-67f85cf4f1df?ixlib=rb-1.2.1&amp;ixid=eyJhcHBfaWQiOjEyMDd9&amp;auto=format&amp;fit=crop&amp;w=300&amp;q=140" transition="fade-transition" class="white--text align-end">
    <!---->
    <v-card-title-stub>

    </v-card-title-stub>
  </v-img-stub>
  <v-card-actions-stub class="map-card__actions">
    <v-subheader-stub>Home.HomeCardMode.undefined</v-subheader-stub>
    <v-spacer-stub></v-spacer-stub>
    <homecarddialog-stub data="[object Object]"></homecarddialog-stub>
  </v-card-actions-stub>
</v-card-stub>
`;
